[FX] Fix default tensor values

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65187

Previously, default values of Tensor type would break in FX codegen, because those values would be Tensor constants that are inserted as attributes on the Module. This makes it so that, when we have a parameter with a default tensor value, we instead emit a sentinel value as the default value, then emit a statement in the function body that will replace the variable with its default value if the sentinel is present

Differential Revision: [D31009105](https://our.internmc.facebook.com/intern/diff/D31009105)

Commit:
[FX] Fix default tensor values

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31009105).

Chillee commented:
I'm a bit worried about this solution.

One thing is that this introduces more things into our IR.

For example, one thing that fails with this approach is

```
def f(a, b = torch.ones(3)):
    return a + b

fx.symbolic_trace(fx.symbolic_trace(f))
```
It fails since we're encountering a new object type in the IR that proxy.py doesn't know how to handle. 

To be honest, I can't come up with any solution I like... At the very least, we just need to support tracing it in Proxy.py

Commit:
Update on "[FX] Fix default tensor values"


Previously, default values of Tensor type would break in FX codegen, because those values would be Tensor constants that are inserted as attributes on the Module. This makes it so that, when we have a parameter with a default tensor value, we instead emit a sentinel value as the default value, then emit a statement in the function body that will replace the variable with its default value if the sentinel is present

Differential Revision: [D31009105](https://our.internmc.facebook.com/intern/diff/D31009105)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31009105).

Commit:
Update on "[FX] Fix default tensor values"


Previously, default values of Tensor type would break in FX codegen, because those values would be Tensor constants that are inserted as attributes on the Module. This makes it so that, when we have a parameter with a default tensor value, we instead emit a sentinel value as the default value, then emit a statement in the function body that will replace the variable with its default value if the sentinel is present

Differential Revision: [D31009105](https://our.internmc.facebook.com/intern/diff/D31009105)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31009105).

Commit:
Update on "[FX] Fix default tensor values"


Previously, default values of Tensor type would break in FX codegen, because those values would be Tensor constants that are inserted as attributes on the Module. This makes it so that, when we have a parameter with a default tensor value, we instead emit a sentinel value as the default value, then emit a statement in the function body that will replace the variable with its default value if the sentinel is present

Differential Revision: [D31009105](https://our.internmc.facebook.com/intern/diff/D31009105)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31009105).

Commit:
Update on "[FX] Fix default tensor values"


Previously, default values of Tensor type would break in FX codegen, because those values would be Tensor constants that are inserted as attributes on the Module. This makes it so that, when we have a parameter with a default tensor value, we instead emit a sentinel value as the default value, then emit a statement in the function body that will replace the variable with its default value if the sentinel is present

Differential Revision: [D31009105](https://our.internmc.facebook.com/intern/diff/D31009105)

[ghstack-poisoned]

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -303,15 +302,15 @@ def _gen_python_code(self, nodes, root_module: str, namespace: _Namespace) -> Py
             Returns: the global name that should be used to reference 'obj' in generated source.
             """
-            if _is_from_torch(obj) and obj != torch.device:  # to support registering torch.device
+            if not is_wrapped and _is_from_torch(obj) and obj != torch.device:  # to support registering torch.device
```
@suo I hacked your hack, please take a look

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31009105).

Commit:
Update on "[FX] Fix default tensor values"


Previously, default values of Tensor type would break in FX codegen, because those values would be Tensor constants that are inserted as attributes on the Module. This makes it so that, when we have a parameter with a default tensor value, we instead emit a sentinel value as the default value, then emit a statement in the function body that will replace the variable with its default value if the sentinel is present

Differential Revision: [D31009105](https://our.internmc.facebook.com/intern/diff/D31009105)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31009105).

Commit:
Update on "[FX] Fix default tensor values"


Previously, default values of Tensor type would break in FX codegen, because those values would be Tensor constants that are inserted as attributes on the Module. This makes it so that, when we have a parameter with a default tensor value, we instead emit a sentinel value as the default value, then emit a statement in the function body that will replace the variable with its default value if the sentinel is present

Differential Revision: [D31009105](https://our.internmc.facebook.com/intern/diff/D31009105)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31009105).

Commit:
Update on "[FX] Fix default tensor values"


Previously, default values of Tensor type would break in FX codegen, because those values would be Tensor constants that are inserted as attributes on the Module. This makes it so that, when we have a parameter with a default tensor value, we instead emit a sentinel value as the default value, then emit a statement in the function body that will replace the variable with its default value if the sentinel is present

Differential Revision: [D31009105](https://our.internmc.facebook.com/intern/diff/D31009105)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31009105).

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -563,13 +568,56 @@ def code(self) -> str:
+        for node in self.graph.nodes:
+            if node.op == 'placeholder' and len(node.args):
+                if isinstance(node.args[0], torch.Tensor):
+                    # HACK: hoisting this logic out of Tracer. We should probably figure
```
Filed issue

Chillee commented on torch/fx/graph_module.py:
```diff
@@ -563,13 +568,56 @@ def code(self) -> str:
         called after editing the contained ``graph``, otherwise the generated
         code of this ``GraphModule`` will be out of date.
+
+        `convert_tensor_default_args` specifies whether `placeholder` nodes with default
```
Is there any reason this isn't just always on?

jamesr66a replied:
Because we're stable release and behavior change like this should be opt-out. I'm not super concerned because these default params basically didn't work before, but the option should still be available to opt out.

Chillee replied:
I'm just wary of new arguments :P

Chillee reviewed (commented):
Could you give a bit more explanation of how this approach actually works? 🤔 Like, give some examples of what the FX code that we output is.

jamesr66a commented:
@Chillee 

First test case generates this code:

```
def forward(self, x, extra = torch.fx.node._DefaultTensorSentinel):
    _tensor_constant0 = self._tensor_constant0
    null_coalesce_tensor_sentinel = torch.fx._symbolic_trace.null_coalesce_tensor_sentinel(extra, _tensor_constant0);  extra = _tensor_constant0 = None
    cat = torch.cat([x, null_coalesce_tensor_sentinel]);  x = null_coalesce_tensor_sentinel = None
    return cat
```

`_DefaultTensorSentinel` is a value to indicate that the user did not specify anything for this `kwarg` (and it needs to be a value like this because a user could legitimately specify something like `None`). This object is traceable as an arg and serializable. `null_coalesce_tensor_sentinel` is `x if x is not _DefaultTensorSentinel else default_value` but is a traceable and serializable function

Chillee reviewed (approved):
LGTM.

One last thought I had - is it plausible to do this purely as an extension of the CodeGen class? Could be potentially nicer if possible.

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -432,7 +431,7 @@ def emit_node(node : Node):
 
                 qualified_name = _get_qualified_name(node.target)
-                global_name = add_global(qualified_name, node.target)
+                global_name = add_global(qualified_name, node.target, node.meta.get('is_wrapped', False))
```
https://github.com/pytorch/pytorch/blob/dae7ed179f60b478d85d9faab06e0e0b42817d96/torch/fx/_symbolic_trace.py#L630

Commit:
Update on "[FX] Fix default tensor values"


Previously, default values of Tensor type would break in FX codegen, because those values would be Tensor constants that are inserted as attributes on the Module. This makes it so that, when we have a parameter with a default tensor value, we instead emit a sentinel value as the default value, then emit a statement in the function body that will replace the variable with its default value if the sentinel is present

Differential Revision: [D31009105](https://our.internmc.facebook.com/intern/diff/D31009105)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31009105).
